db/view/view_building_worker.hh

```diff
@@ -0,0 +1,175 @@
+
+namespace replica {
+class database;
+}
+
+namespace netw {
+class messaging_service;
+}
+
+namespace service {
+class raft_group0_client;
+}
+
+namespace db {
+
+class system_keyspace;
+
+namespace view {
+
+class view_update_generator;
+
+// The view_building_worker is a service responsible for executing
+// tasks scheduled by view_building_coordinator.
+// The service observes `view_building_state` and handles `work_on_view_building_tasks()` RPC.
+// In response to the RPC, the worker start the tasks (check view_building_state.hh)
+// and executes them until they are finished or aborted.
+// If the RPC connection is broken (for example, because the raft leader was changed
+// and current view building coordinator was aborted), the work is continued in background
+// and the new coordinator will attach itself to the ongoing tasks.
+class view_building_worker : public seastar::peering_sharded_service<view_building_worker>, public service::migration_listener::only_view_notifications {
+    /*
+     * View building coordinator may send multiple tasks at once to be executed together.
+     * This is possible only if all of the tasks:
+     * - are of the same type
+     * - their base_id is the same
+     * - their last_token is the same
+     * - their shard_id is the same
+     *
+     * Moreover, any of the task can be aborted during execution. When this happens:
+     * - the batch is not aborted unless it was the last alive task
+     * - entry is removed from _tasks_ids
+     *
+     * When `work` future is finished, it means all tasks in `tasks_ids` are done.
+     *
+     * The batch lives on shard 0 but it might execute its work on a different shard.
+     */
+
+    enum class batch_state {
+        idle,
+        in_progress,
+        finished,
+    };
+
+    class batch {
+    public:
+        batch_state state = batch_state::idle;
+        table_id base_id;
+        locator::tablet_replica replica;
+        std::unordered_map<utils::UUID, view_building_task> tasks;
+
+        shared_future<> work;
+        condition_variable batch_done_cv;
+        sharded<abort_source> abort_sources;
+
+        batch(view_building_worker& vbw, std::unordered_map<utils::UUID, view_building_task> tasks, table_id base_id, locator::tablet_replica replica);
+        future<> start();
+        future<> abort_task(utils::UUID id);
+        future<> abort();
+
+    private:
+        view_building_worker& _vbw;
+
+        future<> do_work();
+        future<> do_build_range(view_building_worker& local_vbw);
+        future<> do_process_staging(view_building_worker& local_vbw);
+    };
+
+    friend class batch;
+
+    struct local_state {
+        std::optional<table_id> processing_base_table = std::nullopt;
+        // Stores ids of views for which the flush was done.
+        // When a new view is created, we need to flush the base table again,
+        // as data might be inserted.
+        std::unordered_set<table_id> flushed_views;
```

Comment:
That's not really a list of flushed views - we didn't flush the views, we only flushed their corresponding base table. The name is confusing, perhaps should be improved in a follow-up.